File: backend/src/sandbox/api/tool/command.py
"""Internal implementation for Phase 3T command-session verbs."""

from __future__ import annotations

from collections.abc import Mapping
from typing import Any

from audit.base import AuditSink
from sandbox.api.tool._daemon_response_parsing import (
    daemon_request_identity_fields,
    parse_timing_map_field,
)
from sandbox.api.tool._operation_audit import run_audited_operation
from sandbox.api.timeouts import shell_dispatch_timeout
from sandbox.api.transport import (
    DAEMON_OP_COMMAND_CANCEL,
    DAEMON_OP_COMMAND_COLLECT_COMPLETED,
    DAEMON_OP_COMMAND_WRITE_STDIN,
    DAEMON_OP_EXEC_COMMAND,
    SandboxTransport,
    call_sandbox_daemon,
)
from sandbox._shared.clock import monotonic_now
from sandbox._shared.models import (
    CommandOutput,
    CommandSessionCancelRequest,
    CommandSessionWriteRequest,
    ExecCommandRequest,
    ExecCommandResult,
)


async def exec_command(
    sandbox_id: str,
    request: ExecCommandRequest,
    *,
    audit_sink: AuditSink | None = None,
    transport: SandboxTransport | None = None,
) -> ExecCommandResult:
    """Run or start a managed command session."""
    total_start = monotonic_now()

    async def _call() -> ExecCommandResult:
        payload = daemon_request_identity_fields(request) | {
            "cmd": request.cmd,
        }
        if request.yield_time_ms is not None:
            payload["yield_time_ms"] = request.yield_time_ms
        if request.timeout is not None:
            payload["timeout"] = request.timeout
        if request.max_output_tokens is not None:
            payload["max_output_tokens"] = request.max_output_tokens
        response = await call_sandbox_daemon(
            sandbox_id,
            DAEMON_OP_EXEC_COMMAND,
            payload,
            timeout=shell_dispatch_timeout(request.timeout),
            transport=transport,
        )
        timings = parse_timing_map_field(response.get("timings"))
        timings["api.exec_command.dispatch_total_s"] = monotonic_now() - total_start
        return _parse_exec_command_result(response, timings=timings)

    return await run_audited_operation(
        audit_sink=audit_sink,
        sandbox_id=sandbox_id,
        operation="exec_command",
        caller=request.caller,
        payload={},
        call=_call,
    )


async def write_stdin(
    sandbox_id: str,
    request: CommandSessionWriteRequest,
    *,
    transport: SandboxTransport | None = None,
) -> ExecCommandResult:
    payload = daemon_request_identity_fields(request) | {
        "command_session_id": request.command_session_id,
        "chars": request.chars,
    }
    if request.yield_time_ms is not None:
        payload["yield_time_ms"] = request.yield_time_ms
    if request.max_output_tokens is not None:
        payload["max_output_tokens"] = request.max_output_tokens
    response = await call_sandbox_daemon(
        sandbox_id,
        DAEMON_OP_COMMAND_WRITE_STDIN,
        payload,
        timeout=shell_dispatch_timeout(None),
        transport=transport,
    )
    return _parse_exec_command_result(response)


async def cancel_command_session(
    sandbox_id: str,
    request: CommandSessionCancelRequest,
    *,
    transport: SandboxTransport | None = None,
) -> ExecCommandResult:
    payload = daemon_request_identity_fields(request) | {
        "command_session_id": request.command_session_id,
    }
    response = await call_sandbox_daemon(
        sandbox_id,
        DAEMON_OP_COMMAND_CANCEL,
        payload,
        timeout=shell_dispatch_timeout(None),
        transport=transport,
    )
    return _parse_exec_command_result(response)


async def collect_command_completions(
    sandbox_id: str,
    *,
    agent_id: str,
    command_session_ids: list[str],
    transport: SandboxTransport | None = None,
) -> list[dict[str, Any]]:
    response = await call_sandbox_daemon(
        sandbox_id,
        DAEMON_OP_COMMAND_COLLECT_COMPLETED,
        {
            "agent_id": agent_id,
            "command_session_ids": command_session_ids,
        },
        timeout=shell_dispatch_timeout(None),
        transport=transport,
    )
    raw = response.get("completions")
    return [dict(item) for item in raw if isinstance(item, Mapping)] if isinstance(raw, list) else []


def _parse_exec_command_result(
    response: Mapping[str, Any],
    *,
    timings: dict[str, float] | None = None,
) -> ExecCommandResult:
    output_raw = response.get("output")
    output = output_raw if isinstance(output_raw, Mapping) else {}
    exit_code_raw = response.get("exit_code")
    exit_code = int(exit_code_raw) if isinstance(exit_code_raw, int) else None
    changed_paths_raw = response.get("changed_paths")
    changed_path_kinds_raw = response.get("changed_path_kinds")
    conflict_reason_raw = response.get("conflict_reason")
    mutation_source_raw = response.get("mutation_source")
    return ExecCommandResult(
        success=str(response.get("status") or "") not in {"error", "timed_out"},
        status=str(response.get("status") or "error"),
        exit_code=exit_code,
        output=CommandOutput(
            stdout=str(output.get("stdout") or ""),
            stderr=str(output.get("stderr") or ""),
        ),
        command_session_id=(
            str(response.get("command_session_id"))
            if response.get("command_session_id")
            else None
        ),
        timings=timings or parse_timing_map_field(response.get("timings")),
        conflict_reason=(
            str(conflict_reason_raw) if conflict_reason_raw is not None else None
        ),
        changed_paths=(
            [str(path) for path in changed_paths_raw]
            if isinstance(changed_paths_raw, list)
            else []
        ),
        changed_path_kinds=(
            {str(key): str(value) for key, value in changed_path_kinds_raw.items()}
            if isinstance(changed_path_kinds_raw, Mapping)
            else {}
        ),
        mutation_source=(
            str(mutation_source_raw) if mutation_source_raw is not None else ""
        ),
        error=response.get("error") if isinstance(response.get("error"), dict) else None,
    )


__all__ = [
    "cancel_command_session",
    "collect_command_completions",
    "exec_command",
    "write_stdin",
]

File: agent-core/crates/eos-sandbox-api/src/tool_api/command.rs
//! Pure command-session helpers: `exec_command`, `exec_stdin`/`write_stdin`,
//! `cancel_command_session`, and `collect_command_completions`. The first three
//! return [`ExecCommandResult`]; `collect_command_completions` returns the raw
//! completion maps (the only verb without a typed result struct).

use eos_types::{JsonObject, SandboxId};
use serde_json::Value;

use crate::error::SandboxApiError;
use crate::models::{
    CommandSessionCancelRequest, CommandSessionWriteRequest, ExecCommandRequest, ExecCommandResult,
    ExecStdinRequest,
};
use crate::ops::DaemonOp;
use crate::timeouts::exec_dispatch_timeout;
use crate::tool_api::parse::{daemon_request_identity_fields, parse_exec_command_result};
use crate::transport::SandboxTransport;

/// Run or start a managed command session.
pub async fn exec_command(
    transport: &dyn SandboxTransport,
    sandbox_id: &SandboxId,
    request: &ExecCommandRequest,
) -> Result<ExecCommandResult, SandboxApiError> {
    let mut payload = daemon_request_identity_fields(&request.base);
    payload.insert("cmd".to_owned(), Value::String(request.cmd.clone()));
    if let Some(yield_time_ms) = request.yield_time_ms {
        payload.insert("yield_time_ms".to_owned(), Value::from(yield_time_ms));
    }
    if let Some(timeout) = request.timeout {
        payload.insert("timeout".to_owned(), Value::from(timeout));
    }
    if let Some(max_output_tokens) = request.max_output_tokens {
        payload.insert(
            "max_output_tokens".to_owned(),
            Value::from(max_output_tokens),
        );
    }
    let response = transport
        .call(
            sandbox_id,
            DaemonOp::ExecCommand,
            payload,
            exec_dispatch_timeout(request.timeout),
        )
        .await?;
    parse_exec_command_result(&response)
}

/// Write characters (stdin) to an open command session.
pub async fn exec_stdin(
    transport: &dyn SandboxTransport,
    sandbox_id: &SandboxId,
    request: &ExecStdinRequest,
) -> Result<ExecCommandResult, SandboxApiError> {
    let mut payload = daemon_request_identity_fields(&request.base);
    payload.insert(
        "command_session_id".to_owned(),
        Value::String(request.command_session_id.clone()),
    );
    payload.insert("chars".to_owned(), Value::String(request.chars.clone()));
    if let Some(yield_time_ms) = request.yield_time_ms {
        payload.insert("yield_time_ms".to_owned(), Value::from(yield_time_ms));
    }
    if let Some(max_output_tokens) = request.max_output_tokens {
        payload.insert(
            "max_output_tokens".to_owned(),
            Value::from(max_output_tokens),
        );
    }
    if request.terminate {
        payload.insert("terminate".to_owned(), Value::Bool(true));
    }
    let response = transport
        .call(
            sandbox_id,
            DaemonOp::ExecStdin,
            payload,
            exec_dispatch_timeout(None),
        )
        .await?;
    parse_exec_command_result(&response)
}

/// Model-facing alias for [`exec_stdin`].
pub async fn write_stdin(
    transport: &dyn SandboxTransport,
    sandbox_id: &SandboxId,
    request: &CommandSessionWriteRequest,
) -> Result<ExecCommandResult, SandboxApiError> {
    exec_stdin(transport, sandbox_id, request).await
}

/// Cancel an open command session.
pub async fn cancel_command_session(
    transport: &dyn SandboxTransport,
    sandbox_id: &SandboxId,
    request: &CommandSessionCancelRequest,
) -> Result<ExecCommandResult, SandboxApiError> {
    let mut payload = daemon_request_identity_fields(&request.base);
    payload.insert(
        "command_session_id".to_owned(),
        Value::String(request.command_session_id.clone()),
    );
    let response = transport
        .call(
            sandbox_id,
            DaemonOp::CommandCancel,
            payload,
            exec_dispatch_timeout(None),
        )
        .await?;
    parse_exec_command_result(&response)
}

/// Collect completed background command sessions for one agent. Returns the raw
/// completion maps (objects only; non-object entries are dropped).
pub async fn collect_command_completions(
    transport: &dyn SandboxTransport,
    sandbox_id: &SandboxId,
    agent_id: &str,
    command_session_ids: &[String],
) -> Result<Vec<JsonObject>, SandboxApiError> {
    let mut payload = JsonObject::new();
    payload.insert("agent_id".to_owned(), Value::String(agent_id.to_owned()));
    payload.insert(
        "command_session_ids".to_owned(),
        Value::Array(
            command_session_ids
                .iter()
                .map(|id| Value::String(id.clone()))
                .collect(),
        ),
    );
    let response = transport
        .call(
            sandbox_id,
            DaemonOp::CommandCollectCompleted,
            payload,
            exec_dispatch_timeout(None),
        )
        .await?;
    let completions = match response.get("completions") {
        Some(Value::Array(items)) => items
            .iter()
            .filter_map(|item| item.as_object().cloned())
            .collect(),
        _ => Vec::new(),
    };
    Ok(completions)
}
